### PR TITLE
add support for enrolledKeys option when secure boot is enabled

### PR DESCRIPTION
**Pull request by nianyush**
Signed-off-by: Nianyu Shen <xiaoyu9964@gmail.com>

### What this PR does
Before this PR:

After this PR:

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note

```

### CODE REVIEW

File: pkg/virt-launcher/virtwrap/manager.go [hunk 1]
```diff
@@ -951,16 +951,17 @@ func (l *LibvirtDomainManager) generateConverterContext(vmi *v1.VirtualMachineIn
 	var efiConf *converter.EFIConfiguration
 	if vmi.IsBootloaderEFI() {
 		secureBoot := vmi.Spec.Domain.Firmware.Bootloader.EFI.SecureBoot == nil || *vmi.Spec.Domain.Firmware.Bootloader.EFI.SecureBoot
+		enrolledKeys := vmi.Spec.Domain.Firmware.Bootloader.EFI.EnrolledKeys == nil || *vmi.Spec.Domain.Firmware.Bootloader.EFI.EnrolledKeys
```
Comment (victortoso) [inline pkg/virt-launcher/virtwrap/manager.go:954]: The matrix of possibilities here are troublesome. This would allow for all three options described in [secureboot docs](https://libvirt.org/kbase/secureboot.html).

If we are going this route we should probably switch to firmware auto-selection which might require a bigger refactor (e.g: let libvirt handle the keys) and have a well defined use cases that we want to support (we will have problems with secureboot in different archs and even in different OS versions).

@andreabolognani, I should have recorded that 1h+ discussion. Would you mind taking a look and see if my note here is valid or I misremember ?

Comment (nianyush) [reply pkg/virt-launcher/virtwrap/manager.go:954]: Yes I would agree that using firmware autoselection seems to be a better idea. My workaround in the original issue was already trying to use that.

Comment (victortoso) [reply pkg/virt-launcher/virtwrap/manager.go:954]: And what problems you've faced when using firmware autoselection?

**Comment by andreabolognani** [comment]
@victortoso the problem with Secure Boot support is that it's effectively a tristate: disabled, enabled but with no keys enrolled, enabled and with keys enrolled. However, since all hardware that ships with Secure Boot support will come with the Microsoft keys already enrolled, from the user's point of view "no keys enrolled" is effectively interpreted as "disabled".

Additionally, depending on the host OS you might need to change your strategy when it comes to run an unsigned OS: some distributions (RHEL) don't ship a build of edk2 without Secure Boot support, while others (Alpine IIRC) don't ship with one that contains it. And there is no good way to know in advance which one will work.

This issue is inherent in the way Secure Boot is implemented. Some time ago I wondered how to expose a user-friendly interface from virt-install (https://github.com/virt-manager/virt-manager/issues/495) and, while I haven't necessarily reached a conclusion, my current take is that it will need to be exposed as a tristate.

I've been itching to go back and actually implement that feature, so maybe there's an opportunity here to land on a shared "language" for both virt-manger and KubeVirt? It'd be great if that could be the case.

Of course I support 100% the idea of dropping KubeVirt's homegrown firmware handling code in favor of the libvirt implementation. But then again I'm quite biased, having spent so much time improving the latter... ;)

File: pkg/virt-launcher/virtwrap/manager.go (hunk 1 above)
Comment (nianyush) [reply pkg/virt-launcher/virtwrap/manager.go:954]: You cannot specify loader and nvram along with firmware autoselect

**Comment by nianyush** [comment]
@andreabolognani Actually the famous microsoft 2011 key is going to be expired in 2026 which will be in the near future, and from what I know is that many hw/os vendors are starting to creating their own keys to sign their fw/os. And especially if you want to have something more secure, e.g have your whole os image signed, doing full disk encryption... you will often find yourself the need to have you own keys enrolled in the system.

> Additionally, depending on the host OS you might need to change your strategy when it comes to run an unsigned OS: some distributions (RHEL) don't ship a build of edk2 without Secure Boot support, while others (Alpine IIRC) don't ship with one that contains it. And there is no good way to know in advance which one will work.
Can you explain this part a little bit? My understanding is that in kubevirt, ed2k ovmf is part of the kubevirt image which means it doesn't really matter if host OS has those packages installed or not.

File: pkg/virt-launcher/virtwrap/manager.go (hunk 1 above)
Comment (andreabolognani) [reply pkg/virt-launcher/virtwrap/manager.go:954]: That was true in the past, but it's not the case anymore.

The change was introduced with [this series](https://lists.libvirt.org/archives/list/devel@lists.libvirt.org/thread/22MYBS3LUZIBS2ZJ4XUGVYY6VQDYQQB6/), which is part of libvirt 9.2.0. Some follow-up fixes were included in libvirt 9.7.0.

These days, the only constraint is that you can't provide information that disagree with each other, for example

```xml
<os firmware='efi'>
  <firmware>
    <feature enabled='yes' name='secure-boot'/>
  </firmware>
  <loader format='qcow2'>/usr/share/edk2/ovmf/OVMF_CODE_4M.qcow2</loader>
</os>
```

This configuration will be rejected because libvirt knows that the specific build of edk2 referenced in the `<loader>` element doesn't have Secure Boot support, which is explicitly requested in the `<firmware>` element.

The following configuration, on the other hand, is perfectly valid:

```xml
<os firmware='efi'>
  <firmware>
    <feature enabled='no' name='enrolled-keys'/>
    <feature enabled='yes' name='secure-boot'/>
  </firmware>
  <loader type='pflash' format='qcow2'>/usr/share/edk2/ovmf/OVMF_CODE_4M.secboot.qcow2</loader>
  <nvram template='/usr/share/edk2/ovmf/OVMF_VARS_4M.qcow2' format='qcow2'/>
</os>
```

In fact, if you provide just

```xml
<os firmware='efi'>
  <firmware>
    <feature enabled='no' name='enrolled-keys'/>
    <feature enabled='yes' name='secure-boot'/>
  </firmware>
</os>
```

or just 

```xml
<os>
  <loader type='pflash' format='qcow2'>/usr/share/edk2/ovmf/OVMF_CODE_4M.secboot.qcow2</loader>
  <nvram template='/usr/share/edk2/ovmf/OVMF_VARS_4M.qcow2' format='qcow2'/>
</os>
```

libvirt will fill in the blanks and the end result will be full snippet seen above in both cases.

Comment (andreabolognani) [reply pkg/virt-launcher/virtwrap/manager.go:954]: There's one important caveat that I failed to mention. The mechanism I just described only works on the condition that the user-provided paths are **the same** as the ones contained in the corresponding firmware descriptor. This is how libvirt can match the two.

Right now KubeVirt hardcodes legacy paths such as `/usr/share/OVMF/OVMF_CODE.fd`, which can't be matched with any firmware descriptor. So none of the above will work, and you're right that in that specific context using the two mechanisms together is not possible.

Incidentally, this also means that KubeVirt does not benefit from the improvements made somewhat recently in the firmware area, such as the switch to 4MB builds and the use of qcow2 format, which results in significant memory usage improvements for aarch64 VMs.

**Comment by andreabolognani** [comment]
> Actually the famous microsoft 2011 key is going to be expired in 2026 which will be in the near future, and from what I know is that many hw/os vendors are starting to creating their own keys to sign their fw/os.

Didn't they create a new key? I expect that future edk2 builds will simply have both enrolled, if that's not the case already.

Things are obviously a bit different in the VM world, but 90% of real hardware will keep coming out with Windows preinstalled, which means that distros are forced to have their releases signed by Microsoft. Requiring the user to mess with enrolled certificates before being able to install an OS would be too high a bar.

> And especially if you want to have something more secure, e.g have your whole os image signed, doing full disk encryption... you will often find yourself the need to have you own keys enrolled in the system.

This is a perfectly valid use case, albeit an advanced one. Most people will be perfectly happy running an off-the-shelf image.

> > Additionally, depending on the host OS you might need to change your strategy when it comes to run an unsigned OS: some distributions (RHEL) don't ship a build of edk2 without Secure Boot support, while others (Alpine IIRC) don't ship with one that contains it. And there is no good way to know in advance which one will work.
> 
> Can you explain this part a little bit? My understanding is that in kubevirt, ed2k ovmf is part of the kubevirt image which means it doesn't really matter if host OS has those packages installed or not.

That's true for the upstream build of KubeVirt, but downstreams will want to make their own container images from their own base OS plus their own QEMU and libvirt packages.

Admittedly it's not very likely that Alpine will start producing its own downstream KubeVirt build any time soon :) and those parties that might want to do that, such as SUSE or Canonical, are probably following similar policies as Red Hat already when it comes to building and signing firmware.

In other words, what I wrote above is technically accurate but might not be a real concern in practice.

**Comment by nianyush** [comment]
> > Actually the famous microsoft 2011 key is going to be expired in 2026 which will be in the near future, and from what I know is that many hw/os vendors are starting to creating their own keys to sign their fw/os.
> 
> Didn't they create a new key? I expect that future edk2 builds will simply have both enrolled, if that's not the case already.
> 
> Things are obviously a bit different in the VM world, but 90% of real hardware will keep coming out with Windows preinstalled, which means that distros are forced to have their releases signed by Microsoft. Requiring the user to mess with enrolled certificates before being able to install an OS would be too high a bar.
> 
> > And especially if you want to have something more secure, e.g have your whole os image signed, doing full disk encryption... you will often find yourself the need to have you own keys enrolled in the system.
> 
> This is a perfectly valid use case, albeit an advanced one. Most people will be perfectly happy running an off-the-shelf image.
> 
> > > Additionally, depending on the host OS you might need to change your strategy when it comes to run an unsigned OS: some distributions (RHEL) don't ship a build of edk2 without Secure Boot support, while others (Alpine IIRC) don't ship with one that contains it. And there is no good way to know in advance which one will work.
> > 
> > 
> > Can you explain this part a little bit? My understanding is that in kubevirt, ed2k ovmf is part of the kubevirt image which means it doesn't really matter if host OS has those packages installed or not.
> 
> That's true for the upstream build of KubeVirt, but downstreams will want to make their own container images from their own base OS plus their own QEMU and libvirt packages.
> 
> Admittedly it's not very likely that Alpine will start producing its own downstream KubeVirt build any time soon :) and those parties that might want to do that, such as SUSE or Canonical, are probably following similar policies as Red Hat already when it comes to building and signing firmware.
> 
> In other words, what I wrote above is technically accurate but might not be a real concern in practice.

Thanks for you detailed explanation! That certainly answers my question and it's new for me to know that downstreames would produce their own image. s
